texar/torch/modules/decoders/decoder_base.py

```diff
@@ -405,8 +406,25 @@ def dynamic_decode(self, helper: Helper, inputs: Optional[torch.Tensor],
 
         while (not torch.all(finished).item() and
                (max_decoding_length is None or time < max_decoding_length)):
-            (next_outputs, decoder_state, next_inputs,
-             decoder_finished) = self.step(helper, time, step_inputs, state)
+
+            next_outputs, decoder_state = \
+                self.step(helper, time, step_inputs, state)
+
+            if max_decoding_length is not None and \
+                    time + 1 == max_decoding_length:
+                # Maximum decoding length reached, mark all batches as finished.
+                # This requires special handling because performing lookup on
+                # position embeddings with `time + 1` may result in IndexError.
+                decoder_finished = torch.tensor(1, dtype=torch_bool,
+                                                device=finished.device)
+                # Since `next_inputs` will not be used, simply create a null
+                # tensor.
+                next_inputs = torch.empty(0)
+            else:
+                logits = next_outputs.logits  # type: ignore
+                sample_id = next_outputs.sample_id  # type: ignore
+                decoder_finished, next_inputs = helper.next_inputs(
```

Comment:
It'd be better to add a `next_inputs` method to the decoder and call here (instead of calling helper's `next_inputs` directly). 

One reason is that in TF, `dynamic_decode` (https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/contrib/seq2seq/python/ops/decoder.py#L268-L472) does not have access to helper, but only decoder. Though we will also update TF `dynamic_decode` to add the `if.. else..` block like here, it's good to not change the factorization in TF (helper -- decoder -- `dynamic_decode`, where `dynamic_decode` does not operate helper directly). To make TF/Pytorch have the similar structures, let's add `next_inputs` to the decoders.